File: tree/lowestCommonAncestor.js
// 寻找节点p和节点q最近的父节点
function lowestCommonAncestor(root, p, q) {
    if(root === null) return null
    // root为p,q中的一个
    if(root === p || root === q) return root
    //分解为求左子树的子问题和右子树的子问题,注意是后序遍历
    let left_have = lowestCommonAncestor(root.left,p,q)    
    let right_have = lowestCommonAncestor(root.right,p,q)
    // p，q分别为root的左右子树
    if(left_have && right_have) return root
    else return left_have ? left_have : right_have
}

File: tree/lowestCommonAncestor.go
package main

// 找出二叉树中某两个节点的第一个共同祖先
func lowestCommonAncestor(root *TreeNode, p *TreeNode, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val == p.Val || root.Val == q.Val {
		// 找到p，q结点
		return root
	}
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)

	// x 的左孩子和右孩子包含 p , q
	if left != nil && right != nil {
		return root
	}
	// x恰好为 p 或 q，右子树包含了另一个节点
	if left != nil {
		return left
	}
	// x恰好为 p 或 q，左子树包含了另一个节点
	return right
}
